webapp/app/.server/service/l2-cancellations.ts

```diff
@@ -234,3 +274,44 @@ export function getL2GovernorAddress(proposalType: L2CancellationType) {
   }
   return l2GovernorAddress;
 }
+
+export async function getAndUpdateL2CancellationByExternalId(
+  externalId: Hex
+): Promise<InferSelectModel<typeof l2CancellationsTable>> {
+  const proposal = await getL2CancellationByExternalId(externalId);
+  if (!proposal) {
+    throw notFound();
+  }
+
+  return await upgradeCancellationStatus(proposal, await getL2VetoNonce());
+}
+
+async function upgradeCancellationStatus(
+  cancellation: InferSelectModel<typeof l2CancellationsTable>,
+  currentNonce: number
+): Promise<InferSelectModel<typeof l2CancellationsTable>> {
+  if (cancellation.status !== l2CancellationStatusEnum.enum.ACTIVE) {
+    return cancellation;
+  }
+
+  if (hexSchema.safeParse(cancellation.transactionHash).success) {
+    cancellation.status = l2CancellationStatusEnum.enum.DONE;
+    await updateL2Cancellation(cancellation.id, cancellation);
+    return cancellation;
+  }
+
+  const state = await getL2ProposalState(cancellation.type, cancellation.externalId);
+  if (!isValidCancellationState(state)) {
+    cancellation.status = l2CancellationStatusEnum.enum.L2_PROPOSAL_EXPIRED;
+    await updateL2Cancellation(cancellation.id, cancellation);
+    return cancellation;
+  }
+
+  if (cancellation.nonce < currentNonce) {
+    cancellation.status = l2CancellationStatusEnum.enum.NONCE_TOO_LOW;
+    await updateL2Cancellation(cancellation.id, cancellation);
+    return cancellation;
+  }
+
+  return cancellation;
```

Comment:
Should we ever reach this line? If not, then I would throw an internal error, so that we don't miss a silent error if any of the business logic ever changes.

Reply:
We agreed on leaving this for next release.